tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java

```diff
@@ -127,15 +122,22 @@ public static void main(String[] args) throws Exception {
             int currentTransactionSize = 0;
             long transactionStartTime = 0;
             for (long i = 0; i < numRecords; i++) {
+
+                if (payloadFilePath != null) {
+                    payload = payloadByteList.get(random.nextInt(payloadByteList.size()));
+                } else if (recordSize != null) {
+                    payload = new byte[recordSize];
```

Comment:
I guess by allocating a new `byte[]` for each record and with big enough `recordSize` we'd generate quite a lot of garbage, so it might be better to allocate `payload` once outside the loop and just fill the buffer here.

Reply:
> generate quite a lot of garbage

really? why it can generate a lot of garbage? you can put some gc log.

> allocate payload once outside the loop 

the payload always same.

Reply:
I'm suggesting to just move the `new byte[recordSize]` outside the loop. Obviously the point of the PR is to _populate_ it inside the loop. If the `new byte[recordSize]` is inside the loop then with, e.g. a 1M record size and large `--num-records` there will be more GC pressure (unless and until the JIT figures out it can reuse the same memory).

Reply:
@tombentley, the result are same.

![image](https://user-images.githubusercontent.com/20113411/114195359-f9c27100-9982-11eb-9103-119a43f5ff7c.png)

Reply:
@tombentley thanks for your review, has updated the patch.

Reply:
Well there _is_ a small difference in the number of young garbage collections, which is what we'd exepect, right? But the ygct is small enough that it makes no difference in a tool which is measuring times using `System.currentTimeMillis()`. That being the case perhaps sticking with the original version would be better, since it would allow you to factor out the payload generation into a separate method.

Reply:
ok, I think you can test it locally.